### ISSUE TITLE
ssl connection to Google Cloud SQL (MySQL) not working

### ISSUE THREAD

**Issue by oulenz** [serge-rider/dbeaver#740]
I am trying to switch from MySQL Workbench to DBeaver, but I can't get the ssl connection to work, as I keep getting an 'access denied' error. I am able to
- connect to the server with the same ssl credentials in MySQL Workbench
- connect without ssl credentials (through user name / password) in DBeaver.

Are there any settings that I might have to change in DBeaver that are automatically set right in MySQL Workbench? Anything particular to Google Cloud SQL? Is it possible DBeaver isn't using ssl even though I checked both `Use SSL` and `Require SSL`?

**Comment by serge-rider** [comment]
Please provide more details.
- Do you use SSH?
- Do you specify SSL certificates?
- What exactly error occurs?
- Check debug log (https://github.com/serge-rider/dbeaver/wiki/Log-files) for any related logs/errors

**Comment by oulenz** [comment]
No SSH. Yes, I specify the certificates. Like I said, these certificates work with MySQL Workbench. The error is:

`java.sql.SQLException: Access denied for user 'root'@'[ip address]' (using password: NO)`

(`or using password: YES` when I specify the password, but that shouldn't matter since I want a ssl connection.

**Comment by serge-rider** [comment]
Hm.
Just a guess: if you are using the latest MySQL driver (6.0.4) then try to use older MySQL driver version (5.x, e.g. 5.1.39). You can change driver version in driver settings->Download/Update->Click on version column.

**Comment by oulenz** [comment]
That didn't help. :-/ Dbeaver-debug.log says i.a:

```
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:569)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:537)
	at com.mysql.cj.jdbc.exceptions.SQLError.createSQLException(SQLError.java:527)
	at com.mysql.cj.jdbc.exceptions.SQLExceptionsMapping.translateException(SQLExceptionsMapping.java:115)
	at com.mysql.cj.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:1634)
	at com.mysql.cj.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:637)
	at com.mysql.cj.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:351)
	at com.mysql.cj.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:224)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:162)
```

**Comment by serge-rider** [comment]
This is a log from MySQL driver 6.x.
They changed whole class structure. I'm thinking about limiting MySQL drivers with 5.x version in DBeaver for a while. BTW #573 is also result of errors in 6.x.

**Comment by oulenz** [comment]
Yes that's from before. For 5.1.39 I get the following stack trace:

```
2016-09-16 14:30:51.341 - Open [http://central.maven.org/maven2/mysql/mysql-connector-java/maven-metadata.xml]
2016-09-16 14:30:59.263 - Connect with 'jdbc:mysql://[server ip address]:[port]/' (mysql5-[hexadecimal code])
2016-09-16 14:30:59.839 - Connection failed (mysql5-[hexadecimal code])
2016-09-16 14:30:59.840 - org.jkiss.dbeaver.model.exec.DBCConnectException: Access denied for user 'root'@'[client ip address]' (using password: YES)
org.jkiss.dbeaver.model.exec.DBCConnectException: Access denied for user 'root'@'[client ip address]' (using password: YES)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:176)
	at org.jkiss.dbeaver.ext.mysql.model.MySQLDataSource.openConnection(MySQLDataSource.java:376)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.connect(JDBCExecutionContext.java:73)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:97)
	at org.jkiss.dbeaver.ext.mysql.model.MySQLDataSource.<init>(MySQLDataSource.java:82)
	at org.jkiss.dbeaver.ext.mysql.MySQLDataSourceProvider.openDataSource(MySQLDataSourceProvider.java:120)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:644)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:74)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionWizard$ConnectionTester.run(ConnectionWizard.java:227)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:103)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: java.sql.SQLException: Access denied for user 'root'@'[client ip address]' (using password: YES)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:963)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3966)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3902)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:875)
	at com.mysql.jdbc.MysqlIO.proceedHandshakeWithPluggableAuthentication(MysqlIO.java:1712)
	at com.mysql.jdbc.MysqlIO.doHandshake(MysqlIO.java:1228)
	at com.mysql.jdbc.ConnectionImpl.coreConnect(ConnectionImpl.java:2253)
	at com.mysql.jdbc.ConnectionImpl.connectOneTryOnly(ConnectionImpl.java:2284)
	at com.mysql.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:2083)
	at com.mysql.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:806)
	at com.mysql.jdbc.JDBC4Connection.<init>(JDBC4Connection.java:47)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:404)
	at com.mysql.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:410)
	at com.mysql.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:328)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:162)
	... 10 more
```

**Comment by serge-rider** [comment]
That's better.
- Are you sure user root'@'[client ip address] has connect permissions? (I see that MySQL workbench works but maybe it is run from another machine/user).
- Does "Require SSL" option (or any other SSL-related options) changes anything?

**Comment by oulenz** [comment]
I have MySQL Workbench open in another window, so same address. (There's no issue with using the same certificate simultaneously, right?) This also occurs if we set 0.0.0.0/0 as an allowed network, i.e. all networks.

(De)checking `Require SSL` or any of the other boxes in the SSL tab doesn't change anything.

I'm on Ubuntu btw.

Since you labelled this `can't reproduce`, I assume you can successfully connect to a MySQL server that requires ssl? If so, I'm wondering whether this might be particular to Google Cloud SQL.

**Comment by serge-rider** [comment]
Yes, I can connect with SSL in my local environment. But there are tons of different SSL-related configs on both server and client side.
Will try it with another hosting a bit later..

**Comment by oulenz** [comment]
I've been able to reproduce my problem with new ssl certificates on Windows, with both the old and the new MySQL driver. I.e. it's still not working, while I have no issues with MySQL Workbench.

(But MySQL Workbench crashes so often it's not funny, so I'd really like to switch.)

**Comment by twastvedt** [comment]
I'm seeing the exact same issue. Windows 7 64 bit.

**Comment by serge-rider** [comment]
Still can't reproduce that. Looks like Google Cloud SSL configuration differs from standard (which I use for testing).
Version 3.7.5 has improved debug log - it includes additional SSL info dump. 
Please check dbeaver-debug.log file for any any useful information.

**Comment by oulenz** [comment]
Tried again with 3.7.6, both with versions 5.1.39 and 5.1.40 of the driver:

```
2016-10-11 12:08:39.099 - Connect with 'jdbc:mysql://[server ip address]:[port]/' (mysql5-[hexadecimal code])
2016-10-11 12:08:39.693 - Connection failed (mysql5-[hexadecimal code])
2016-10-11 12:08:39.696 - org.jkiss.dbeaver.model.exec.DBCConnectException: Access denied for user 'root'@'80.91.33.151' (using password: YES)
org.jkiss.dbeaver.model.exec.DBCConnectException: Access denied for user 'root'@'[client ip address]' (using password: YES)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:176)
	at org.jkiss.dbeaver.ext.mysql.model.MySQLDataSource.openConnection(MySQLDataSource.java:376)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.connect(JDBCExecutionContext.java:73)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:97)
	at org.jkiss.dbeaver.ext.mysql.model.MySQLDataSource.<init>(MySQLDataSource.java:82)
	at org.jkiss.dbeaver.ext.mysql.MySQLDataSourceProvider.openDataSource(MySQLDataSourceProvider.java:120)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:644)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:74)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionWizard$ConnectionTester.run(ConnectionWizard.java:223)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:103)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: java.sql.SQLException: Access denied for user 'root'@'80.91.33.151' (using password: YES)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:963)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3966)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3902)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:875)
	at com.mysql.jdbc.MysqlIO.proceedHandshakeWithPluggableAuthentication(MysqlIO.java:1712)
	at com.mysql.jdbc.MysqlIO.doHandshake(MysqlIO.java:1228)
	at com.mysql.jdbc.ConnectionImpl.coreConnect(ConnectionImpl.java:2253)
	at com.mysql.jdbc.ConnectionImpl.connectOneTryOnly(ConnectionImpl.java:2284)
	at com.mysql.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:2083)
	at com.mysql.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:806)
	at com.mysql.jdbc.JDBC4Connection.<init>(JDBC4Connection.java:47)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:404)
	at com.mysql.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:410)
	at com.mysql.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:328)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:162)
	... 10 more
2016-10-11 12:08:53.097 - Open [http://central.maven.org/maven2/mysql/mysql-connector-java/maven-metadata.xml]
2016-10-11 12:08:57.809 - Open [http://central.maven.org/maven2/mysql/mysql-connector-java/5.1.40/mysql-connector-java-5.1.40.pom]
2016-10-11 12:09:01.462 - Open [http://central.maven.org/maven2/mysql/mysql-connector-java/5.1.40/mysql-connector-java-5.1.40.jar]
2016-10-11 12:09:09.592 - Connect with 'jdbc:mysql://[server ip address]:[port]/' (mysql5-[hexadecimal code])
2016-10-11 12:09:10.136 - Connection failed (mysql5-[hexadecimal code])
2016-10-11 12:09:10.138 - org.jkiss.dbeaver.model.exec.DBCConnectException: Access denied for user 'root'@'80.91.33.151' (using password: YES)
org.jkiss.dbeaver.model.exec.DBCConnectException: Access denied for user 'root'@'[client ip address]' (using password: YES)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:176)
	at org.jkiss.dbeaver.ext.mysql.model.MySQLDataSource.openConnection(MySQLDataSource.java:376)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCExecutionContext.connect(JDBCExecutionContext.java:73)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.<init>(JDBCDataSource.java:97)
	at org.jkiss.dbeaver.ext.mysql.model.MySQLDataSource.<init>(MySQLDataSource.java:82)
	at org.jkiss.dbeaver.ext.mysql.MySQLDataSourceProvider.openDataSource(MySQLDataSourceProvider.java:120)
	at org.jkiss.dbeaver.registry.DataSourceDescriptor.connect(DataSourceDescriptor.java:644)
	at org.jkiss.dbeaver.runtime.jobs.ConnectJob.run(ConnectJob.java:74)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionWizard$ConnectionTester.run(ConnectionWizard.java:223)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:103)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:55)
Caused by: java.sql.SQLException: Access denied for user 'root'@'80.91.33.151' (using password: YES)
	at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:964)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3970)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:3906)
	at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:873)
	at com.mysql.jdbc.MysqlIO.proceedHandshakeWithPluggableAuthentication(MysqlIO.java:1710)
	at com.mysql.jdbc.MysqlIO.doHandshake(MysqlIO.java:1226)
	at com.mysql.jdbc.ConnectionImpl.coreConnect(ConnectionImpl.java:2253)
	at com.mysql.jdbc.ConnectionImpl.connectOneTryOnly(ConnectionImpl.java:2284)
	at com.mysql.jdbc.ConnectionImpl.createNewIO(ConnectionImpl.java:2083)
	at com.mysql.jdbc.ConnectionImpl.<init>(ConnectionImpl.java:806)
	at com.mysql.jdbc.JDBC4Connection.<init>(JDBC4Connection.java:47)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at com.mysql.jdbc.Util.handleNewInstance(Util.java:425)
	at com.mysql.jdbc.ConnectionImpl.getInstance(ConnectionImpl.java:410)
	at com.mysql.jdbc.NonRegisteringDriver.connect(NonRegisteringDriver.java:328)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSource.openConnection(JDBCDataSource.java:162)
	... 10 more
```

**Comment by serge-rider** [comment]
Doesn't look like SSL connection (no SSL handshake logs).
Enable "Require SSL" option before testing.

**Comment by oulenz** [comment]
I did that. It is as I observed in my original question, as if these settings aren't being used at all. :-/
 
![image](https://cloud.githubusercontent.com/assets/13745168/19312225/c87b5d9a-9091-11e6-8ea5-500d1052bb92.png)

**Comment by oulenz** [comment]
Tried it again in 3.8.2, same issue.

**Comment by serge-rider** [comment]
Eventually I had a chance to test Google Cloud.
First of all it uses SSL certificate private keys.
Seconds of all private key is in PKCS1 format (instead of typical PKCS8).

Anyhow - now DBeaver supports this. Fix will be in version 4.

**Comment by oulenz** [comment]
Thank you very much for taking the trouble of testing this out yourself! The support you provide for DBeaver is really commendable!

**Comment by ronjouch** [comment]
@serge-rider hi! I got DBeaver to connect to a Google Cloud SQL / PostgreSQL instance, but I have two remaining questions. Suppose the `psql` command passed by Google is

```
psql "sslmode=verify-full sslrootcert=server-ca.pem \
      sslcert=client-cert.pem sslkey=client-key.pem \
      hostaddr=123.456.789.1 \
      host=accountname-12345:instancename \
      port=5432 \
      user=postgres dbname=postgres"
```

I was able to connect in DBeaver, but:

1. Passing the `hostaddr` IP above as `Host` in DBeaver's Connection Settings screen. What's this other `host` (`accountname-12345:instancename`) above? It doesn't seem necessary for a successful connection, is it something Google specific I can safely ignore?
2. In the `SSL` tab of the `Network` settings, I wasn't able to find a DBeaver equivalent to `sslmode=verify-full sslrootcert=server-ca.pem`. If I pass this `server-ca.pem` as `Root Certificate`, the field gets cleared on save, and the only `SSL mode` I got to work with `verify-full` is `org.postgresql.ssl.NonValidatingFactory`. Should I worry about it? How should I configure this properly?  
![dbeaver-uncertain-sslmode-settings](https://user-images.githubusercontent.com/522085/28976549-55822238-790c-11e7-99f0-f019ef623f5f.png)

Thanks for the continued work on DBeaver 👍👍👍.
